File: LinkedList/linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def traverse_linked_list(self):
        current = self.head
        while current:
            print(current.data, end=" ")
            current = current.next
        print()

    def insert_at_start(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def insert_at_index(self, index, data):
        new_node = Node(data)
        if index == 0:
            new_node.next = self.head
            self.head = new_node
            return
        current = self.head
        i = 0
        while current and i < index - 1:
            current = current.next
            i += 1
        if current is None:
            print("Invalid index!")
            return
        new_node.next = current.next
        current.next = new_node

    def delete_at_index(self, index):
        if self.head is None:
            print("List is empty!")
            return
        if index == 0:
            self.head = self.head.next
            return
        current = self.head
        i = 0
        while current and i < index - 1:
            current = current.next
            i += 1
        if current is None or current.next is None:
            print("Invalid index!")
            return
        current.next = current.next.next

linked_list = LinkedList()
linked_list.insert_at_end(1)
linked_list.insert_at_end(2)
linked_list.insert_at_end(3)
linked_list.insert_at_start(0)
linked_list.insert_at_index(2, 5)
linked_list.delete_at_index(1)
linked_list.traverse_linked_list()

File: LinkedList/linkedlist.kt
class Node(var data: Int) {
    var next: Node? = null
}

class LinkedList {
    var head: Node? = null

    fun traverseLinkedList() {
        var current = head
        while (current != null) {
            print("${current.data} ")
            current = current.next
        }
        println()
    }

    fun insertAtStart(data: Int) {
        val newNode = Node(data)
        newNode.next = head
        head = newNode
    }

    fun insertAtEnd(data: Int) {
        val newNode = Node(data)
        if (head == null) {
            head = newNode
            return
        }
        var current = head
        while (current!!.next != null) {
            current = current.next
        }
        current.next = newNode
    }

    fun insertAtIndex(index: Int, data: Int) {
        val newNode = Node(data)
        if (index == 0) {
            newNode.next = head
            head = newNode
            return
        }
        var current = head
        var i = 0
        while (current != null && i < index - 1) {
            current = current.next
            i++
        }
        if (current == null) {
            println("Invalid index!")
            return
        }
        newNode.next = current.next
        current.next = newNode
    }

    fun deleteAtIndex(index: Int) {
        if (head == null) {
            println("List is empty!")
            return
        }
        if (index == 0) {
            head = head!!.next
            return
        }
        var current = head
        var i = 0
        while (current != null && i < index - 1) {
            current = current.next
            i++
        }
        if (current == null || current.next == null) {
            println("Invalid index!")
            return
        }
        current.next = current.next!!.next
    }
}

fun main() {
    val linkedList = LinkedList()
    linkedList.insertAtEnd(1)
    linkedList.insertAtEnd(2)
    linkedList.insertAtEnd(3)
    linkedList.insertAtStart(0)
    linkedList.insertAtIndex(2, 5)
    linkedList.deleteAtIndex(1)
    linkedList.traverseLinkedList()
}
